### PR TITLE
Add/edit/remove database connection and ingest data into project

### PR DESCRIPTION
**Pull request by AstridKery**
This is a rather gigantic MR, but it basically does the following:

* allow admin users to create database connections, either when creating the project, or updating it
* allow admin users to edit or delete those database connections in the project update page
* add a button to the details page to pull the data from the database on demand if connected
* Allow the user to pull the database instead of uploading a file when creating project

NOTE: I'm purposefully leaving out database export and setting up database ingest schedules for future work, to keep this MR as manageable as possible.

I welcome any and all code critiques and of course if you find any bugs that's good too. One thing I'd love from each of you would be a review of the user workflow. What do you like about it? What is confusing? What were you expecting to happen that didn't?

Thanks!

### CODE REVIEW

**Comment by dsteedRTI** [comment]
I tried importing data from a database connection on project upload and encountered this error:
Screen Shot 2022-02-03 at 10 51 29 AM
My workflow was:

- Go through normal steps adding project
- Added a database connection to the test credentials from Ethan
- Instead of doing a file upload for data, I selected to import data from the database

When I tried to upload the data and finish, it gave a loading indicator for a few seconds then redirected me to that above error page.

**Comment by AstridKery** [comment]
> I tried importing data from a database connection on project upload and encountered this error: Screen Shot 2022-02-03 at 10 51 29 AM My workflow was:
> 
> * Go through normal steps adding project
> * Added a database connection to the test credentials from Ethan
> * Instead of doing a file upload for data, I selected to import data from the database
> 
> When I tried to upload the data and finish, it gave a loading indicator for a few seconds then redirected me to that above error page.

I assumed the folder was being created automatically but it was not. This should be fixed now.

File: backend/django/core/forms.py [hunk 1]
```diff
@@ -418,3 +457,71 @@ def clean_data(self):
...
+    port = forms.IntegerField(required=False)
+    dbname = forms.CharField(initial="", required=False, max_length=50)
+    driver = forms.CharField(
+        initial="ODBC Driver 17 for SQL Server", required=False, max_length=200
+    )
+
+    def __init__(self, *args, **kwargs):
+        super(ExternalDatabaseWizardForm, self).__init__(*args, **kwargs)
+
+    def clean(self):
```
Comment (peridoteagle) [inline backend/django/core/forms.py:489]: I like this setup

File: backend/django/core/forms.py [hunk 2]
```diff
@@ -418,3 +457,71 @@ def clean_data(self):
...
+                    if self.cleaned_data.get(field) in ["", None]:
+                        self._errors[field] = self.error_class(
+                            ["This field is required for a MS SQL database connection."]
+                        )
+                        field_error = True
+            if field_error:
+                raise ValidationError("Please fix field errors before resubmitting.")
+
+            engine_database = get_connection(db_type, self.cleaned_data)
+            test_connection(
```
Comment (peridoteagle) [inline backend/django/core/forms.py:521]: I also like the test of the database connection and think it is helpful. I made a typo when I was first inputting the connection information and the error helped me figure it out

File: backend/django/core/migrations/0066_alter_externaldatabase_project.py [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+import django.db.models.deletion
+from django.db import migrations, models
+
+
+class Migration(migrations.Migration):
+
+    dependencies = [
+        ("core", "0065_alter_externaldatabase_project"),
+    ]
+
```
Comment (peridoteagle) [inline backend/django/core/migrations/0066_alter_externaldatabase_project.py:12]: I'll admit that I'm not super familiar with the Migration piece and wasn't entirely sure if there was anything specific I should look for here when reviewing. If so, feel free to let me know

File: backend/django/core/templates/projects/create/create_wizard_external_db.html [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+
+{% load i18n %}
+{% load static %}
+{% load render_bundle from webpack_loader %}
+
+{% block head %}
+{{ wizard.form.media }}
+{{ form.media.css }}
+{% endblock %}
+
```
Comment (peridoteagle) [inline backend/django/core/templates/projects/create/create_wizard_external_db.html:12]: This looks good

File: backend/django/core/templates/projects/update/external_db.html [hunk 5]
```diff
@@ -0,0 +1,119 @@
+{% extends "base.html" %}
+{% block page_title %}}Update Project Database Connection{% endblock %}
```
Comment (peridoteagle) [inline backend/django/core/templates/projects/update/external_db.html:2]: Small thing, but there a bracket } shows up in the page title

File: backend/django/core/templates/projects/update/external_db.html [hunk 6]
```diff
@@ -0,0 +1,119 @@
...
+        if ($(this).prop('checked') == true) {
+            sql_fields.show();
+            database_fields.show();
+        } else {
+            sql_fields.hide();
+            database_fields.hide();
+        }
+    });
+
+    $('#submitButton').click(function (e) {
```
Comment (peridoteagle) [inline backend/django/core/templates/projects/update/external_db.html:108]: Tested this and it worked well

File: backend/django/core/views/api.py [hunk 7]
```diff
@@ -130,3 +139,72 @@ def download_model(request, project_pk):
...
+        request: The POST request
+        project_pk: Primary key of the project
+    Returns:
+        {}
+    """
+    response = {}
+    profile = request.user.profile
+    project = Project.objects.get(pk=project_pk)
+
+    # Make sure coder is an admin
```
Comment (peridoteagle) [inline backend/django/core/views/api.py:159]: I like the admin requirement

**Review by peridoteagle — APPROVED** [review]
This looks super solid to me. Really nice work! It'd be good for us to review the current behavior of what happens when a database connection is updated or deleted with Ethan to make sure he is on the same page.

File: backend/django/core/migrations/0066_alter_externaldatabase_project.py (hunk 3 above)
Comment (AstridKery) [reply backend/django/core/migrations/0066_alter_externaldatabase_project.py:12]: The migrations are automatically generated. So I guess the main thing to review is really the models.py file and the new tables/fields defined in it

File: backend/django/core/templates/projects/update/external_db.html (hunk 5 above)
Comment (AstridKery) [reply backend/django/core/templates/projects/update/external_db.html]: Great catch! Just updated

**Pull request merged**